cmd/main.go

```diff
@@ -4,137 +4,151 @@ import (
-	prioritySet := make(map[int]bool)
-	for name, upstream := range cfg.Upstreams {
-		if upstream.Type != "azure" && upstream.Type != "openai" {
-			fmt.Printf("Invalid type for upstream %s: %s\n", name, upstream.Type)
-			return
-		}
-		if upstream.Type == "azure" && upstream.URL == "" {
-			fmt.Printf("Missing URL for Azure upstream %s\n", name)
-			return
-		}
-		if upstream.Type == "openai" && upstream.URL != "" {
-			fmt.Printf("URL should not be provided for OpenAI upstream %s\n", name)
-			return
-		}
-		if upstream.Priority <= 0 {
-			fmt.Printf("Invalid priority for upstream %s: %d\n", name, upstream.Priority)
-			return
-		}
-		if _, exists := prioritySet[upstream.Priority]; exists {
-			fmt.Printf("Duplicate priority for upstream %s: %d\n", name, upstream.Priority)
-			return
-		}
-		prioritySet[upstream.Priority] = true
-	}
 
-	// If listeners are provided via CLI, override the ones from the config
 	if cliListeners != "" {
-		cfg.Listeners = []internal.Listener{}
-		for _, cliListener := range strings.Split(cliListeners, ",") {
-			parts := strings.Split(cliListener, ":")
-			if len(parts) != 2 {
-				fmt.Println("Invalid listener format, skipping:", cliListener)
-				continue
-			}
-			cfg.Listeners = append(cfg.Listeners, internal.Listener{Interface: parts[0], Port: parts[1]})
-		}
+		overrideListeners(cfg, cliListeners)
 	}
 
-	// Initialize logger
-	logger, err := internal.InitializeLogger(cfg.LogLevel)
+	// Initialize the logger in logger.go
+	logger, err := internal.InitializeLogger(logLevel)
 	if err != nil {
-		panic("Couldn't initialize logger")
+		log.Fatal("Couldn't initialize logger: ", err)
 	}
-	defer logger.Sync()
+	ctx := createContext(cfg, logger)
 
-	// Create context with logger and config
-	ctx := context.WithValue(context.Background(), "logger", logger)
-	ctx = context.WithValue(ctx, "config", cfg)
-	logger.Debug("Added config to context: ", zap.Any("config", ctx.Value("config")))
-
-	// Register HTTP handler
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
 		internal.Response(ctx, w, r)
 	})
 
-	// Get hostname
-	hostname, err := os.Hostname()
-	if err != nil {
-		logger.Error("Could not get hostname", zap.Error(err))
+	startListeners(ctx, cfg)
+}
+
+func overrideListeners(cfg *internal.Config, cliListeners string) {
+	const ExpectedParts = 2
+
+	cfg.Listeners = []internal.Listener{}
+
+	for _, cliListener := range strings.Split(cliListeners, ",") {
+		parts := strings.Split(cliListener, ":")
+		isValidFormat := len(parts) == ExpectedParts
+
+		if !isValidFormat {
+			log.Error("Invalid listener format, skipping: ", cliListener)
+
+			continue
+		}
+
+		newListener := internal.Listener{Interface: parts[0], Port: parts[1]}
+		cfg.Listeners = append(cfg.Listeners, newListener)
+	}
+}
+
+func createContext(cfg *internal.Config, logger *log.Logger) context.Context {
+	ctx := context.Background()
+	ctx = context.WithValue(ctx, "logger", logger)
+	ctx = context.WithValue(ctx, "config", cfg)
```

Comment:
Two problems here.
1. We do not use contexts to pass around configuration.
2. Your context key should be not only a unique 'string' but also a unique type.
```suggestion
	type MyCoolContextType string
	var configKey MyCoolContextType = "config"
	ctx = context.WithValue(ctx, configKey, cfg)
```